src/appleseed.python/bindblendermeshconverter.cpp

```diff
@@ -0,0 +1,219 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// appleseed.renderer headers.
+#include "renderer/api/object.h"
+
+// appleseed.foundation headers.
+#include "foundation/platform/python.h"
+
+namespace bpy = boost::python;
+using namespace renderer;
+
+// Work around a regression in Visual Studio 2015 Update 3.
+#if defined(_MSC_VER) && _MSC_VER == 1900
+namespace boost
+{
+}
+#endif
+
+// Blender data structures
+struct MeshFace
+{
+    unsigned int v[4];
+    short mat_nr;
+    char edcode, flag;
+};
+
+struct MeshVert
+{
+    float co[3];
+    short no[3];
+    char flag, bweight;
+};
+
+struct MeshTexFace
+{
+    float uv[4][2];
+    void *tpage;
+    char flag, transp;
+    short mode, tile, unwrap;
+};
+
+
+// The following function takes a series of pointers to Blender mesh data and modifies the appleseed MeshObject entity
+//      blender_object = the appleseed MeshObject created earlier in the export process
+//      vertices_length = the numbe rof vertices in the mesh
+//      vertex_pointer = a string pointer to the first element of the vertex array
+//      Same template applies to mesh faces
+
+void convert_mesh(
+    MeshObject*                 blender_mesh,
+    const size_t                vertices_length,
+    const size_t                vertex_pointer,
+    const size_t                triangle_length,
+    const size_t                triangle_pointer,
+    const size_t                uv_layer_pointer,
+    const bool                  export_normals,
+    const bool                  export_uvs)
+{
+    // Convert size_t numbers to actual pointers
+    MeshVert* vertices = reinterpret_cast<MeshVert *>(vertex_pointer);
+    MeshFace* triangles = reinterpret_cast<MeshFace *>(triangle_pointer);
+    MeshTexFace* uv_face = reinterpret_cast<MeshTexFace *>(uv_layer_pointer);
+
+    blender_mesh->reserve_vertices(vertices_length);
+    blender_mesh->reserve_triangles(triangle_length);
+
+    if (export_normals == true)
+    {
+        blender_mesh->reserve_vertex_normals(vertices_length);
+    }
+
+    if (export_uvs == true)
+    {
+        blender_mesh->reserve_tex_coords(vertices_length);
+    }
+
+    // Push vertices
+    for (size_t vertex_index = 0; vertex_index < vertices_length; ++vertex_index)
+    {
+        MeshVert &vert = vertices[vertex_index];
+        blender_mesh->push_vertex(
+            GVector3(
+                vert.co[0],
+                vert.co[1],
+                vert.co[2]));
+
+        // Push normals
+        if (export_normals == true)
+        {
+            blender_mesh->push_vertex_normal(
+                GVector3(
+                    vert.no[0],
+                    vert.no[1],
+                    vert.no[2]));
+        }
+    }
+
+    int uv_index = 0;
+
+    // Push triangles
+    for (size_t tri_index = 0; tri_index < triangle_length; ++tri_index)
+    {
+        MeshFace &face = triangles[tri_index];
```

Comment:
What if the face is a quad?

Reply:
It can't be.  There is a triangluation step in the mesh translator.

Reply:
appleseed can't handle quads right?

Reply:
```python
def __get_blender_mesh(self, scene, triangulate=True):
        settings = 'RENDER' if self.__export_mode != ProjectExportMode.INTERACTIVE_RENDER else 'PREVIEW'
        me = self.bl_obj.to_mesh(
            scene,
            apply_modifiers=True,
            settings=settings,
            calc_tessface=False)

        if triangulate:
            bm = bmesh.new()
            bm.from_mesh(me)
            bmesh.ops.triangulate(bm, faces=bm.faces)
            bm.to_mesh(me)
            bm.free()

        return me
```

Reply:
> appleseed can't handle quads right?

No. We plan to. Embree supports quad meshes and it will be needed for subdivs as well.
I did some work on it, but it's not done yet.

Reply:
The blenderseed side of things can be modified when that is ready.